### PR TITLE
Specialize copy_if for small input sizes

### PR DESCRIPTION
**Pull request by adamfidel**
This PR provides an optimization for the `copy_if` algorithm for sizes that are small enough to fit into a single work-group. This should improve the performance of `copy_if`, `remove`, `remove_if`, `remove_copy`,  and `remove_copy_if`.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:636]: I think `in_rng.begin()` doesn't work in case of pipelined range with our "simple" views like
`***_view_simple`, because ones don't have begin/end operator, just `operator[]`.
We didnot need begin/end before and added simplest implementation for the "views" which provide the necessary semantic.

**Comment by MikeDvorskiy** [comment]
> This PR provides an optimization for the `copy_if` algorithm for sizes that are small enough to fit into a single work-group. This should improve the performance of `copy_if`, `remove`, `remove_if`, `remove_copy`, and `remove_copy_if`.

Am I right that we are introducing two new patterns to our SYCL backend - `parallel_scan_copy `and `parallel_copy_if` ?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:886]: I guess should be  commented what is `monotonic_dispatcher` and why we use it

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:875]: probably, `::std::uint16_t` ?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:625]: As far as I understand usage of "subgroup load/store" is enabling just for such rare case when "n==wg_size", right?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:598]: `::std::uint16_t`?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:605]: `::std::uint32_t`?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:894]: Formatting looks strange to me.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1767]: I'm not sure if this and the following changes improve the readability.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:871]: Would it make this code clearer to specify `__max_slm_size` as half `local_mem_size` and multiply `__req_slm_size` by two?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:790]: Did you experiment with the work group size? A smaller size could suffice to saturate the memory bandwidth but might reduce the complexity of the group scan.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:797]: Similar comment as above about tuning this parameter. Other algorithms have shown best performance with 32 elements per work item while this uses 16 at most, I think.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:576]: I think a comment about what this dispatcher does would help here.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:607]: This algorithm might benefit from using USM host memory for the result similar to the discussions for reduce in #874.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:886]: Will add documentation.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:875]: Good catch, done.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:871]: Yes, that would be better. Will change.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:790]: This size was chosen for scan, but I did not do experiments specifically for `copy_if`. I will experiment a bit with these parameters and get back to you.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:797]: I'll test out different parameters.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:598]: Done.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:605]: Done.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:894]: This is from `clang-format`. Perhaps I should manually adjust it and ignore its suggestions for this line.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1767]: This is from `clang-format`. I will revert the changes for lines that I did not touch.

**Comment by adamfidel** [comment]
> > This PR provides an optimization for the `copy_if` algorithm for sizes that are small enough to fit into a single work-group. This should improve the performance of `copy_if`, `remove`, `remove_if`, `remove_copy`, and `remove_copy_if`.
> 
> Am I right that we are introducing two new patterns to our SYCL backend - `parallel_scan_copy `and `parallel_copy_if` ?

Yes, although  `parallel_scan_copy` is just moving code from `__pattern_scan_copy` in `algorithm_impl_hetero.h`. This is needed to provide a fallback path for when we cannot call the single-work-group copy_if specialization.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:630]: Could we benefit from adding this group load into SLM with the strided access pattern as a utility function somewhere? I could see this being used quite a bit throughout different kernels. The same thing could be done for storing results in global memory from SLM.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:797]: I tested out a few different parameters and they were within 1-2% of each other with no clear winner. Based on those experiments, I think I will just keep it to 1024 for now.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:636]: > guarded only for ranges that support `.begin()`.

I guess it is not enough here... `subgroup.load` does requires a pointer and probably doesn't  work with a general iterator.
Also, can we use `sycl::vec` instead of "subgroup" (intel extension) ?
https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:vector.type

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: hm... Should we whether  a special code branch for so rare case, just a "bench" case ( "n==wg_size") ? 
What if we keep the code which processes "residual"? and If residual is zero, does a compiler get rid of the "residual code"?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:636]: 1. I replaced the condition with one that checks if `begin()` returns a pointer.
2. I don't see how we can use `sycl::vec` here since this code is mainly just trying to pull items from global memory in an efficient way.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:630]: I think the use of the unary operator here makes this code a little specific to this algorithm. The store is even more algorithm-specific as it first checks a predicate first before copying. In the general case, we should be able to use bulk stores.

I think the introduction of group-level loads and stores to SYCL may help us standardize this a bit when those are introduced.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:603]: Despite the comment, I'm confused as to why this and the `static_cast` later on are necessary in this pattern.  

I think I understand `__get_tuple_type<>` generally, its so that we can convert our internal tuple to a `std::tuple`.  It seems to be used for the other users of `__parallel_scan_copy`, but I'm wondering if that is because of their internal use of zip iterator.  Is it still necessary here?  

If so, should we be using this elsewhere?

Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:603]: This issue is exposed in the `zip_iterator.pass` test, which calls `copy_if` with a `zip_iterator` input iterator and a `transform_iterator` output iterator. Without the cast, we are met with the following error:
```
[...]/include/oneapi/dpl/pstl/tuple_impl.h:261:31: error: reference to type 'const std::_Swallow_assign' could not bind to an lvalue of type 'const int'
    __value_holder(_Up&& t) : value(::std::forward<_Up>(t))
```
This is the same workaround used in the `__copy_by_mask` functor for the original `copy_if` algorithm. See the comment at https://github.com/oneapi-src/oneDPL/blob/6bb45531b7bd6908a16897ebd8fedbb18d365430/include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h#L517-L535

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:654]: Done.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h]: Thanks, done.

File: include/oneapi/dpl/pstl/utils.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/utils.h]: Good catch. I fixed it.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:633]: Okay, thanks. I think we can leave it as is then. This is probably something more broad to evaluate on a wider scale in the future.

File: include/oneapi/dpl/pstl/utils_ranges.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/utils_ranges.h]: I see this has been refactored and is no longer relevant.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:603]: Thanks for the additional information.  I think it is fine here in the context of this PR.

I'll make an issue to discuss this `__get_tuple_type<>` workaround in general.  I think it is there to enable a usage pattern which should be discouraged (and can be rewritten to not require the workaround). Also, there are other algorithm(s?) for which a similar usage pattern will cause the same compilation errors that don't currently use this workaround.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (timmiesmith) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:607]: Waiting on #874 before merging this shouldn't be done. We may need to revisit this to update this algorithm though.

Comment (julianmi) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:607]: I agree. #874 is currently blocked by a compiler issue. We will need to evaluate its outcome and investigate where and how USM host memory is beneficial.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: I believe this was discussed in the original single-work-group scan PR:
https://github.com/oneapi-src/oneDPL/pull/783#discussion_r1113201765

I would assume the same applies here in the `copy_if` case.

**Review by danhoeflinger — APPROVED** [review]
This LGTM.  It seems like @MikeDvorskiy may still have some open unresolved questions though, so it would be good to get his OK on those topics before merging.

**Comment by adamfidel** [comment]
> @MikeDvorskiy Am I right that we are introducing two new patterns to our SYCL backend - parallel_scan_copy and parallel_copy_if ?

Yes, this introduces a specific pattern for `__parallel_copy_if`, which is distinct from `__parallel_scan_copy`. With scan_copy, there are higher-order functions that are passed around to implement algorithms such as `unique_copy`. Ideally, we could have added the single-work-group specialization to `__parallel_scan_copy` but it turned out to be difficult to support all of the algorithms that relied on the higher-order function capabilities, so we instead split the pattern into two: one specifically for `copy_if` and the original scan-copy pattern.

File: include/oneapi/dpl/pstl/utils_ranges.h
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/utils_ranges.h:60]: uglification

**Review by MikeDvorskiy — APPROVED** [review]
LGTM with a minor comment

File: include/oneapi/dpl/pstl/utils_ranges.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/utils_ranges.h:60]: Could you please clarify how you would like this line to be formatted? Should I move the inheritance to a new line?

File: include/oneapi/dpl/pstl/utils_ranges.h
Comment (adamfidel) [reply include/oneapi/dpl/pstl/utils_ranges.h:65]: I just applied the suggested change from clang-format. Does that match what you were expecting to see?

File: include/oneapi/dpl/pstl/utils_ranges.h
Comment (akukanov) [reply include/oneapi/dpl/pstl/utils_ranges.h:60]: @adamfidel, uglification means adding undescores in front of all the names, like this:
```
template <class _T, class = void>
struct __range_has_raw_ptr_iterator : ::std::false_type
```

Note also that we typically use `typename` instead of `class` for template parameters.

Comment (adamfidel) [reply include/oneapi/dpl/pstl/utils_ranges.h]: Thanks for the clarification @akukanov. I updated the names to have double underscores and fixed the class/typename issue.

@MikeDvorskiy would you mind re-approving?

**Review by danhoeflinger — APPROVED** [review]
LGTM,  I checked over the uglification changes, and tried to confirm that the merge didn't spoil anything (as well as my previous approval).

**Pull request merged**
